src/plugins/data/public/ui/dataset_selector/index.tsx

```diff
@@ -40,15 +40,12 @@ const ConnectedDatasetSelector = ({
     };
   }, [queryString]);
 
-  const handleDatasetChange = useCallback(
-    (dataset?: Dataset) => {
-      setSelectedDataset(dataset);
-      if (dataset) {
-        const query = queryString.getInitialQueryByDataset(dataset);
-        queryString.setQuery(query);
-        onSubmit!(queryString.getQuery());
-        queryString.getDatasetService().addRecentDataset(dataset);
-      }
+  const onSelect = useCallback(
+    (partialQuery: Partial<Query>) => {
+      const query = queryString.getInitialQuery(partialQuery);
+      setSelectedDataset(query.dataset);
+      queryString.setQuery(query);
```

Comment:
This is not a good pattern. We are effectively trying to keep two separate states in sync. That means we have 2 sources of truth, which is risky